Deployment (#6)
avantifellows/plio-analytics: docs/AUTHENTICATION.md

> ## Authentication

Reviewer: need to set refresh timer to 20 days or something.

> Plio Analytics uses Auth0 authentication to manage access tokens and valid API requests from the client app, which in Plio's ecosystem is the frontend app.

Reviewer 2: ```suggestion
Plio Analytics uses Auth0 authentication to manage access tokens and valid API requests from a client app.
```
I don't think it's needed. Any client app can connect to this analytics service right?
Just an opinion though!

> Plio Analytics uses Auth0 authentication to manage access tokens and valid API requests from the client app, which in Plio's ecosystem is the frontend app.

Reviewer: Are we using Security context?

Author: Right now we'll proceed without security context as we have machine-to-machine API calls between frontend and analytics.

The ACL checks will be handled by frontend/backend repos.

> Plio Analytics uses Auth0 authentication to manage access tokens and valid API requests from the client app, which in Plio's ecosystem is the frontend app.

Reviewer: <img width="711" alt="Screenshot 2021-05-12 at 12 56 49 PM" src="https://user-images.githubusercontent.com/13185699/117935644-8fa24080-b321-11eb-8bc8-b46dbdf73cac.png">
Here the key mentioned is for connecting cubejs to the JWK - and the JWT is the key for each client of cubejs (like our plio backend) right?

> 

Reviewer: Are we using auth0.js or auth sdk spa?

Reviewer: Or we are ignoring it for now as it seems to be important for custom claims? https://cube.dev/docs/security/jwt/auth0#create-an-application-custom-claims

Author: Yeah we don't need custom claims as of now. We're just relying on Auth0 for access token and auth verification at cubejs.

Plio Analytics relies on client credentials API which is suitable for machine-to-machine data exchange without any user consent. This generates access token for the client app instead of per-user access token.

For more details on Auth0 configuration, check out the [official documentation from Cube.js](https://cube.dev/docs/security/jwt/auth0).

### Configurating Auth0
1. Sign up or log into your [Auth0 account](https://auth0.com/).
2. From the left navigation, go to APIs and create a new API.
   1. Name your API.
   2. Set the identifier to your app url. Like `https://myapp.com/`. This must be a unique identifier and not to be reused in other APIs.
   3. Signing Algorithm should be `RS256`.
3. From the left navigation, go to Applications.
   1. You will see an application was created automatically with your api name and `(Test application)`.
   2. Select this app and go to settings tab. Rename by removing the (Test Application) part and save.

> 4. Go to Quick Start tab and copy the example code for "Getting an access token for your API". Run the command in your shell. It should return a valid response with access token, expiry time and token type.

Reviewer: what is to be done with the access token and other values? is this only for testing?

Author: This is only for testing. i'll update the step.

### Configuring CUBE_JS environment variables
Add the following environment variables for this CubeJS project after replacing the `<>` parts with your configs.
```sh
CUBEJS_JWK_URL=https://<AUTH0-SUBDOMAIN>.auth0.com/.well-known/jwks.json
CUBEJS_JWT_AUDIENCE=<APPLICATION_URL>
CUBEJS_JWT_ISSUER=https://<AUTH0-SUBDOMAIN>.auth0.com/
CUBEJS_JWT_ALGS=RS256
CUBEJS_JWT_CLAIMS_NAMESPACE=<CLAIMS_NAMESPACE>
```

#### `CUBEJS_JWK_URL`

> This is the url to fetch the JSON wek keys details. Replace the `<AUTH0-SUBDOMAIN>` by the domain name for your account. The domain can be determined from `Your app > Settings > Domain`.

Reviewer 2: ```suggestion
This is the url to fetch the JSON web keys details. Replace the `<AUTH0-SUBDOMAIN>` by the domain name for your account. The domain can be determined from `Your app > Settings > Domain`.
```

   1. You will see an application was created automatically with your api name and `(Test application)`.
   2. Select this app and go to settings tab. Rename by removing the (Test Application) part and save.
4. To test your set up, go to Quick Start tab and copy the example code for "Getting an access token for your API". Run the command in your shell. It should return a valid response with access token, expiry time and token type.

> 5. Make sure you have updated the Plio backend repository [environment variables](https://github.com/avantifellows/blob/master/docs/ENV.md#auth0-for-plio-analytics) with the Auth0 app settings.

Reviewer: the link seems to be invalid (the repo name is missing)
